presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java

```diff
@@ -0,0 +1,173 @@
+import static io.prestosql.spi.StandardErrorCode.ALREADY_EXISTS;
+import static io.prestosql.spi.StandardErrorCode.INVALID_PROCEDURE_ARGUMENT;
+import static io.prestosql.spi.block.MethodHandleUtil.methodHandle;
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class RegisterPartitionProcedure
+        implements Provider<Procedure>
+{
+    private static final MethodHandle REGISTER_PARTITION = methodHandle(
+            RegisterPartitionProcedure.class,
+            "registerPartition",
+            ConnectorSession.class,
+            String.class,
+            String.class,
+            String.class,
+            List.class,
+            List.class);
+
+    private final Supplier<TransactionalMetadata> hiveMetadataFactory;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final HiveMetastoreClosure metastore;
+
+    @Inject
+    public RegisterPartitionProcedure(Supplier<TransactionalMetadata> hiveMetadataFactory, HiveMetastore metastore, HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.metastore = new HiveMetastoreClosure(requireNonNull(metastore, "metastore is null"));
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "register_partition",
+                ImmutableList.of(
+                        new Procedure.Argument("schema_name", VARCHAR),
+                        new Procedure.Argument("table_name", VARCHAR),
+                        new Procedure.Argument("location", VARCHAR),
+                        new Procedure.Argument("partition_columns", new ArrayType(VARCHAR)),
+                        new Procedure.Argument("partition_values", new ArrayType(VARCHAR))),
+                REGISTER_PARTITION.bindTo(this));
+    }
+
+    public void registerPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doRegisterPartition(session, schemaName, tableName, location, partitionColumnNames, partitionValues);
+        }
+    }
+
+    private void doRegisterPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        HiveIdentity identity = new HiveIdentity(session);
+        HdfsContext hdfsContext = new HdfsContext(session, schemaName, tableName);
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+
+        Path partitionLocation = new Path(location);
```

Comment:
In Hive's `ALTER TABLE ... ADD PARTITION`, `LOCATION` is optional.  I found that to be a great usability factor - I didn't want to worry about typos in specifying the path.  Unfortunately, procedures don't support optional arguments, so you would have to pass in a `null`, which is kind of ugly, but we do in other procedures.  Have we considered not asking the user to specify a location?  We pretty much have the logic for that implemented in `SyncPartitionMetadataProcedure`.

Reply:
Makes sense. I've made PR for optional procedure arguments: https://github.com/prestosql/presto/pull/2706

Reply:
When it got merge I'll make a follow up PR for making location optional (with default)

Reply:
Let's gather more feedback before adding more flexibility. We would want to add the right options that are commonly needed or cannot be done otherwise.